isis/src/viking/apps/vikcal/CalParameters.cpp

```diff
@@ -360,24 +363,44 @@ namespace Isis {
   }
 
   /**
-   * Calculates the distance from the sun at the specified time
+   * Calculates the distance from Mars to the sun at the specified time.
+   * Try to useing the camera assiciated with the cube first, if that
+   * doesn't work fall back to using the SPICE data.
    *
-   * @param t iTime
+   * @param t The UTC time at which the sun distance is being requested
+   * @param iCube The cube we are calibrating
+   *
+   * @return Distance from the Sun to Mars in km
    */
-  void CalParameters::CalcSunDist(QString t) {
-    NaifStatus::CheckErrors();
-    double sunv[3];
-    SpiceDouble lt, et;
-    FileName fname1 = (FileName)"$base/kernels/lsk/naif0007.tls";
-    FileName fname2 = (FileName)"$base/kernels/spk/de405.bsp";
-    QString tempfname1 = fname1.expanded();
-    QString tempfname2 = fname2.expanded();
-    furnsh_c(tempfname1.toLatin1().data());
-    furnsh_c(tempfname2.toLatin1().data());
-    utc2et_c(t.toLatin1().data(), &et);
-    spkezp_c(10, et, "J2000", "LT+S", 499, sunv, &lt);
-    p_dist1 = sqrt(sunv[0] * sunv[0] + sunv[1] * sunv[1] + sunv[2] * sunv[2]);
-    NaifStatus::CheckErrors();
+  double CalParameters::CalcSunDist(QString t, Cube *iCube) {
+    try {
+      Camera *cam;
+      cam = iCube->camera();
+      iTime startTime(t);
+      cam->setTime(startTime);
+      return cam->sunToBodyDist();
+    }
+    catch(IException &e) {
+      // Failed to instantiate a camera, try furnishing kernels directly
+      try {
+        NaifStatus::CheckErrors();
+        double sunv[3];
+        SpiceDouble lt, et;
+        FileName fname1 = (FileName)"$base/kernels/lsk/naif0007.tls";
+        FileName fname2 = (FileName)"$base/kernels/spk/de405.bsp";
+        QString tempfname1 = fname1.expanded();
+        QString tempfname2 = fname2.expanded();
+        furnsh_c(tempfname1.toLatin1().data());
+        furnsh_c(tempfname2.toLatin1().data());
+        utc2et_c(t.toLatin1().data(), &et);
+        spkezp_c(10, et, "J2000", "LT+S", 499, sunv, &lt);
+        return sqrt(sunv[0] * sunv[0] + sunv[1] * sunv[1] + sunv[2] * sunv[2]);
+        NaifStatus::CheckErrors();
+      }
```

Comment:
I understand why we would want to divert to the old method if there isn't enough information on the cube, but in the future would we not want to add the necessary info to the cube?

Reply:
For ISIS cameras, the info is included in the attached SPICE data. It just takes the Camera (techniqualy Spice) a simple calculation to return it. For CSMCameras, the info isn't currently available. @jessemapel and I have talked about how to add it for the CSMCamera